sklearn/ensemble/_hist_gradient_boosting/histogram.pyx

```diff
@@ -317,13 +317,23 @@ cpdef void _subtract_histograms(
         hist_struct [:, ::1] hist_b,  # IN
         hist_struct [:, ::1] out) noexcept nogil:  # OUT
     """compute (hist_a - hist_b) in out"""
+    # Note that subtraction of large sums of floating point numbers, as we have here,
+    # can exhibit catastrophic cancallation. This is in particular true for gradients
+    # as they can be positive and negative, while hessians are are non-negative.
+    # Remember that gradients and hessians are originally computed in
+    # G_H_DTYPE_C = float32 precision. Therefore, if sum_gradients and sum_hessians are
+    # float64, we don't loose precision. But if we would again use float32 for
+    # summation, we would need to take care of floating point errors.
     cdef:
         unsigned int i = 0
     for i in range(n_bins):
         out[feature_idx, i].sum_gradients = (
             hist_a[feature_idx, i].sum_gradients -
             hist_b[feature_idx, i].sum_gradients
         )
+        # Here, we could protect for negative hessians by setting:
+        #     sum_hessians = max(0, sum_hessians)
+        # But as we use float64 for summing float32, that's veeeery unlikely.
```

Comment:
Same here, this comment describes a possible future, but I do not think it is actionable?

Reply:
I tried out to set sum_hessians to float32. All our tests pass!! Only the higgs boson benchmark fails badly. So, I'm not yet 100% convinced that we neeeeeever get a negative hessian. We just never check or test it.